near-sdk/src/lib.rs

```diff
@@ -102,31 +112,182 @@ extern crate quickcheck;
+///
+/// #### [`collections`] module:
+///
+/// ```rust
+/// # use near_sdk_macros::near;
+/// use near_sdk::collections::LookupMap;
+///
+/// #[near(contract_state)]
+/// pub struct StatusMessage {
+///    records: LookupMap<String, String>,
+/// }
+/// ```
+///
+/// * list of [**host functions**](collections#calls-to-host-functions-used-in-implementation) used for [`collections`] implementation
+///
+/// ### Implementation details of `#[near(contract_state)]` macro and **host functions** calls used
+///
+/// If the details of [ABI](https://github.com/near/abi) generation layer are put aside, then the macro performs roughly the following:  
+///
+/// ```rust
+/// # use near_sdk::near;
+/// #[near(contract_state)]
+/// pub struct Contract { /* .. */ }
+/// ```
+///
+/// 1. Macro adds derived implementations of [`borsh::BorshSerialize`]/[`borsh::BorshSerialize`] for `Contract` type
+/// 2. Macro defines a global `CONTRACT_SOURCE_METADATA` variable, which is a string of json serialization of [`near_contract_standards::contract_metadata::ContractSourceMetadata`](https://docs.rs/near-contract-standards/latest/near_contract_standards/contract_metadata/struct.ContractSourceMetadata.html).
+/// 3. Macro defines `contract_source_metadata` function:
+///     ```rust,no_run
+///     #[no_mangle]
+///     pub extern "C" fn contract_source_metadata() { /* .. */ }
+///     ```
+///    which
+///     1. calls [`env::setup_panic_hook`] host function
+///     2. calls [`env::value_return`] host function with bytes of `CONTRACT_SOURCE_METADATA` from step 2.
+///
+/// ##### using [cargo-expand](https://crates.io/crates/cargo-expand) to view actual macro results
+///
+/// The above is an approximate description of what macro performs.
+///
+/// Running the following in a contract's crate is a way to introspect more details of its operation:
+///
+/// ```bash,ignore
+/// cargo expand --lib --target wasm32-unknown-unknown
+/// # this has additional code generated for ABI layer
+/// cargo expand --lib --features near-sdk/__abi-generate
+/// ```
+/// ---
+///
+/// ## `#[near]` (annotates impl blocks)
+///
+/// This macro is used to define the code for view-only and mutating methods for contract types,
+/// annotated by [`#[near(contract_state)]`](near#nearcontract_state-annotates-structsenums).
+///
+/// ### Basic example
+/// ```rust
+/// use near_sdk::{near, log};
+///
+/// # #[near(contract_state)]
+/// # pub struct Contract {
+/// #     greeting: String,
+/// # }
+/// #[near]
+/// impl Contract {
+///     // view method
+///     pub fn get_greeting(&self) -> String {
+///         self.greeting.clone()
+///     }
+///
+///     // mutating method
+///     pub fn set_greeting(&mut self, greeting: String) {
+///         log!("Saving greeting: {greeting}");
+///         self.greeting = greeting;
+///     }
+/// }
+/// ```
+///
+/// ### Implementation details of `#[near]` macro and **host functions** calls used
```

Comment:
To be honest, I would keep impl details separately rather than overload the doc. I think, most users don't really care.

Reply:
And keep a reference to details here.

Reply:
> I think, most users don't really care.

Yes, this one is really more for advanced users and also for new sdk contributors to keep the impression that the core of implementation is really quite simple as opposed to being A.U.T.O.M.A.G.I.C.A.L.

Also this keeps the doc more cohesive and i was only able to understand the Default::default() for state behaviour and .unwrap_or_default() piece while doing this doc snippet.

Reply:
the section also provides a reference point for explaining how other attributes work, so that explanation for other attributes can be made more concise, and only mention what has changed by using an attribute.